File: Strings/1662_Check_If_Two_String_Arrays_Are_Equivalent.py
"""

1662. Check If Two String Arrays are Equivalent
Given two string arrays word1 and word2, return true if the two arrays represent the same string, and false otherwise.
A string is represented by an array if the array elements concatenated in order forms the string.

Example 1:
Input: word1 = ["ab", "c"], word2 = ["a", "bc"]
Output: true

Difficulty : easy

Approach : Two Pointer
1. Initialize two pointers i and j to 0, representing the current index in word1
2. Initialize two pointers p and q to 0, representing the current character index in the current word of word1 and word2 respectively.
3. While both i and j are within the bounds of word1 and word2:
4. Compare the characters word1[i][p] and word2[j][q].
5. If they are not equal, return false.
6. Increment p and q to move to the next character in the current words.
7. If p reaches the end of word1[i], increment i and reset p to 0 to move to the next word in word1.
8. If q reaches the end of word2[j], increment j and reset q to 0 to move to the next word in word2.
9. After the loop, check if both i and j have reached the end of their respective   arrays. If they have, return true; otherwise, return false.
Time Complexity : O(n + m) where n is the total number of characters in word1 and m is the total number of characters in word2.
Space Complexity : O(1) since we are using a constant amount of extra space.

"""
class Solution:
    def arrayStringsAreEqual(self, word1: List[str], word2: List[str]) -> bool:
        #O(n**2) O(n)
        # str1 = ""
        # str2 = ""

        # for word in word1:
        #     str1 += word

        # for word in word2:
        #     str2 += word

        # return str1 == str2

        # Using joins O(n=m)
        # str1 = "".join(word1)
        # str2 = "".join(word2)

        # return str1 == str2

        #Most Optimized O(n + m) O(1)
        i = j = 0  # Word indexes
        p = q = 0  # Character indexes

        while i < len(word1) and j < len(word2):

            if word1[i][p] != word2[j][q]:
                return False

            p += 1
            q += 1

            # Move to next word in word1
            if p == len(word1[i]):
                i += 1
                p = 0

            # Move to next word in word2
            if q == len(word2[j]):
                j += 1
                q = 0

        return i == len(word1) and j == len(word2)

File: Strings/1662_Check_If_Two_String_Arrays_Are_Equivalent.java
/*
1662. Check If Two String Arrays are Equivalent
Given two string arrays word1 and word2, return true if the two arrays represent the same string, and false otherwise.
A string is represented by an array if the array elements concatenated in order forms the string.

Example 1:
Input: word1 = ["ab", "c"], word2 = ["a", "bc"]
Output: true

Difficulty : easy

Approach : Stringbuilder.
1. Concatenate all strings in word1 and word2 using StringBuilder.
2. Compare the resulting strings for equality.
3. Return true if they are equal, otherwise return false.
Time Complexity: O(n + m) where n is the total length of strings in word1 and m is the total length of strings in word2.
Space Complexity: O(n + m) for storing the concatenated strings in StringBuilder.

*/

class Solution {
    public boolean arrayStringsAreEqual(String[] word1, String[] word2) {
        
        // String str1 = "";
        // String str2 = "";

        // for (int i = 0; i < word1.length; i++) {
        //     str1 += word1[i];
        // }

        // for (int i = 0; i < word2.length; i++) {
        //     str2 += word2[i];
        // }

        // return str1.equals(str2);

        //2.Stringbuilder Version
        // StringBuilder str1 = new StringBuilder();
        // StringBuilder str2 = new StringBuilder();

        // for (int i = 0; i < word1.length; i++) {
        //     str1.append(word1[i]);
        // }

        // for (int i = 0; i < word2.length; i++) {
        //     str2.append(word2[i]);
        // }

        // return str1.toString().equals(str2.toString());

        //3.joiner
        // return String.join("", word1).equals(String.join("", word2));

        //4 Pointer

        int i = 0;
        int j = 0;

        int p1 = 0;
        int p2 = 0;

        while (i < word1.length && j < word2.length) {

            if (word1[i].charAt(p1) != word2[j].charAt(p2)) {
                return false;
            }

            p1++;
            p2++;

            // Move to next word in word1
            if (p1 == word1[i].length()) {
                i++;
                p1 = 0;
            }

            // Move to next word in word2
            if (p2 == word2[j].length()) {
                j++;
                p2 = 0;
            }
        }

        return i == word1.length && j == word2.length;
    
    }
}